spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md

```diff
@@ -350,96 +446,79 @@ Access control on counterparty should inform choice of timeout values, i.e. time
-    currentChannel.counterpartyChannelIdentifier, counterpartySequence))
-
-    // get current sequence on this channel
-    currentSequence = provableStore.get(channelUpgradeSequencePath(portIdentifier, channelIdentifier))
-
-    // move upgrade sequence to fresh sequence for new upgrade attempt
-    if channel.state === types.OPEN {
-        channel.upgradeSequence = channel.upgradeSequence + 1
-
+    // if OPEN, then initialize handshake with upgradeFields
+    // otherwise, assert that the upgrade fields are the same for crossing-hellos case
+    if currentChannel.state == OPEN {
         // if the counterparty sequence is greater than the current sequence, we fast forward to the counterparty sequence
         // so that both channel ends are using the same sequence for the current upgrade
+        // initUpgradeChannelHandshake will increment the sequence so after that call
+        // both sides will have the same upgradeSequence
         if counterpartyUpgradeSequence > channel.upgradeSequence {
-            channel.upgradeSequence = counterpartyUpgradeSequence
+            channel.upgradeSequence = counterpartyUpgradeSequence - 1
         }
 
-        // this is first message in upgrade handshake on this chain so we must store original channel in restore channel path
-        // in case we need to restore channel later.
-        privateStore.set(channelRestorePath(portIdentifier, channelIdentifier), currentChannel)
-        provableStore.set(channelPath(portIdentifier, channelIdentifier), channel)
+        initUpgradeChannelHandshake(portIdentifier, channelIdentifier, upgradeFields, counterpartyUpgrade.timeout)
+    } else if currentChannel.state == INITUPGRADE {
+        existingUpgrade = publicStore.get(channelUpgradePath)
+        abortTransactionUnless(existingUpgrade.fields == upgradeFields)
+    }
+
+    // get counterpartyHops for given connection
+    connection = getConnection(currentChannel.connectionIdentifier)
+    counterpartyHops = getCounterpartyHops(connection)
+
+    // construct counterpartyChannel from existing information and provided
+    // counterpartyUpgradeSequence
+    counterpartyChannel = ChannelEnd{
+        state: INITUPGRADE,
+        ordering: currentChannel.ordering,
+        counterpartyPortIdentifier: portIdentifier,
+        counterpartyChannelIdentifier: channelIdentifier,
+        connectionHops: counterpartyHops,
+        version: currentChannel.version,
+        sequence: counterpartyUpgradeSequence,
     }
 
+    // call startFlushUpgrade handshake to move channel from INITUPGRADE to TRYUPGRADE and start flushing
+    // upgrade is blocked on this channelEnd from progressing until flush completes on both ends
+    startFlushUpgradeHandshake(portIdentifier, channelIdentifier, upgradeFields, counterpartyChannel, counterpartyUpgrade, TRYUPGRADE, proofChannel, proofUpgrade, proofHeight)
```

Comment:
at this point in in the code when `startFlushUpgradeHandshake` is called, in the non-crossing hellos scenario our own `channel.State` would still be in `UPGRADEINIT`. wouldn't this mean that the `startFlush...` subprotocol would fail on [its first check? ](https://github.com/cosmos/ibc/pull/967/files#diff-69d7c4ac1d4cdeeabcc757d3a34e85945ebc5613a034a52ebd0769cd8a0fc560R285)

Reply:
ah, nevermind, it's the passed in `ChannelState` param and not our own channel state that is being checked. maybe a clearer naming, something like `NextChannelState`

Reply:
that sounds good to me!